### PR TITLE
[PM-7084] 2/6: Add shared two-factor-auth orchestrator component, and TOTP two-factor component

### PR DESCRIPTION
**Pull request by quexten**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-7084
https://bitwarden.atlassian.net/browse/PM-9098
https://bitwarden.atlassian.net/browse/PM-9024
https://bitwarden.atlassian.net/browse/PM-9056
https://bitwarden.atlassian.net/browse/PM-9057

## 📔 Objective

This PR introduced a feature flagged shared two-factor-auth component along with the first shared authentication method - TOTP. Tests are ported from the old shared two-factor.component.

> [!IMPORTANT]  
> **STACKED PR:** This set of changes is split into a set of smaller PR's that can be reviewed independently, in parallel. Merging still should be done in order once all PR's are reviewed. Merging can be done incrementally as the feature is feature flagged.

#9767 - Shared two-fa options component
#9768 - Shared two-fa orchestrator component & totp authenticator component
#9769 - Shared two-fa yubikey component
#9770 - Shared two-fa email component
#9771 - Shared two-fa webauthn component
#9772 - Shared two-fa duo component

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/web/src/app/oss-routing.module.ts [hunk 1]
```diff
@@ -248,10 +250,9 @@ const routes: Routes = [
         path: "2fa",
         canActivate: [unauthGuardFn()],
         children: [
-          {
+          ...twofactorRefactorSwap(TwoFactorComponent, TwoFactorAuthComponent, {
             path: "",
-            component: TwoFactorComponent,
-          },
+          }),
```
Comment (jlf0dev) [inline apps/web/src/app/oss-routing.module.ts:255]: Question: does this one not need the anon layout? I'm trying to figure out why this one is different from the other routing module changes.

File: libs/common/src/enums/feature-flag.enum.ts [hunk 2]
```diff
@@ -50,6 +52,7 @@ export const DefaultFeatureFlagValue = {
   [FeatureFlag.EmailVerification]: FALSE,
   [FeatureFlag.InlineMenuFieldQualification]: FALSE,
   [FeatureFlag.MemberAccessReport]: FALSE,
+  [FeatureFlag.TwoFactorComponentRefactor]: TRUE,
```
Comment (jlf0dev) [inline libs/common/src/enums/feature-flag.enum.ts:55]: Question: can you expand on why this should be enabled by default? I would think we only want to enable it during testing.

File: apps/desktop/src/auth/two-factor-auth.component.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    FormFieldModule,
+    AsyncActionsModule,
+    RouterLink,
+    CheckboxModule,
+    TwoFactorOptionsComponent,
+    TwoFactorAuthAuthenticatorComponent,
+  ],
+  providers: [I18nPipe],
+})
+export class TwoFactorAuthComponent extends BaseTwoFactorAuthComponent {}
```
Comment (jlf0dev) [inline apps/desktop/src/auth/two-factor-auth.component.ts:41]: Is this used later? Can we just use the lib two factor component?

File: apps/web/src/app/oss-routing.module.ts (hunk 1 above)
Comment (quexten) [reply apps/web/src/app/oss-routing.module.ts:255]: This is already wrapped in an anonlayoutwrapper https://github.com/bitwarden/clients/blob/033c5ee7ecc1c7f1211300c4f7f9ca9efefd4675/apps/web/src/app/oss-routing.module.ts#L175

File: apps/desktop/src/auth/two-factor-auth.component.ts (hunk 3 above)
Comment (quexten) [reply apps/desktop/src/auth/two-factor-auth.component.ts:41]: We need this later to import the platform specific duo and email modules. With just authauthenticatorcomponent we of course do not need it.

**Review by jlf0dev — APPROVED** [review]
👏

**Pull request merged**